Adds jsonld context, Adds support for jsonld typed values

maechler opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

This pull request is just an implementation proposal, it is not yet ready to be merged. I have not put any effort into fixing the tests.

This pull request wants to add json-ld typed values: https://www.w3.org/TR/json-ld/#typed-values
e.g.

**entity** 
```php
    /**
     * @var string | array
     *
     * @ApiProperty(iri="http://xmlns.com/foaf/0.1/firstName");
     *
     */
    private $firstName;

    /**
     * @var string | array
     *
     * @ApiProperty(iri="http://xmlns.com/foaf/0.1/lastName", valueType="literal");
     *
     */
    private $lastName;
```


**context**
```json
{
    "@context": {
        "@vocab": "http://swissbib-hydra.lo/apidoc.jsonld#",
        "hydra": "http://www.w3.org/ns/hydra/core#",
        "firstName": "http://xmlns.com/foaf/0.1/firstName",
        "lastName": {
            "@id": "http://xmlns.com/foaf/0.1/lastName",
            "@type": "literal"
        }
    }
}
```

What do you think?

dunglas commented on src/JsonLd/ContextBuilder.php:
```diff
@@ -107,6 +107,11 @@ public function getResourceContext(string $resourceClass, int $referenceType = U
                     '@id' => $id,
                     '@type' => '@id',
                 ];
+            } else if (null !== $propertyMetadata->getValueType()) {
```
Can be simplified using `elseif`.

dunglas commented on src/Metadata/Property/PropertyMetadata.php:
```diff
@@ -321,4 +323,29 @@ public function withChildInherited(string $childInherited) : self
     }
+
+    /**
+     * JSON-LD value type.
```
We try to make the Metadata component as independent as possible of serialization formats. IMO the value type is OK (it can be used in various serialization formats) but the comment should be made more generic.

dunglas commented:
This method should be updated too: https://github.com/api-platform/core/blob/master/src/Hydra/Serializer/DocumentationNormalizer.php#L282-L319

teohhanhui commented:
What about having these as attributes, so that it can be JSON-LD specific, instead of trying to find a general enough "type" declaration that fits all the formats (sounds like an impossible task).

maechler commented:
I can also move it to the attributes. What would be the convention for the name?
Just use "value_type"="something" ?

teohhanhui commented:
There is already a `hydra_context` key, perhaps we can add `jsonld_context`?

maechler commented:
So it would be something like ?

```php

    /**
     * @var string | array
     *
     * @ApiProperty(iri="http://xmlns.com/foaf/0.1/lastName", attributes={"jsonld_context": {"value_type":"literal"}})
     *
     */
    private $lastName;
```

dunglas commented:
Tyes, it looks better!

teohhanhui commented:
```php
/**
 * @var string
 *
 * @ApiProperty(iri="http://xmlns.com/foaf/0.1/lastName", attributes={
 *     "jsonld_context"={"@type"="literal"},
 * })
 *
 */
private $lastName;
```

maechler commented:
Looks fine for me, I guess that's the way to go:)

teohhanhui commented on src/JsonLd/ContextBuilder.php:
```diff
@@ -97,6 +97,7 @@ public function getResourceContext(string $resourceClass, int $referenceType = U
             }
 
             $convertedName = $this->nameConverter ? $this->nameConverter->normalize($propertyName) : $propertyName;
+            $jsonldContext = $propertyMetadata->getAttributes()['jsonld_context'] ?? [];
```
What's set in the `jsonld_context` attribute should be used as the base. We don't want to limit it to just `@type`.

teohhanhui replied:
See for example how `hydra_context` is handled in https://github.com/api-platform/core/blob/7ac18660fc04f9c261238d447ac52993ff8786da/src/Hydra/Serializer/DocumentationNormalizer.php#L209-L273

maechler replied:
Great input! I did not think of that, I updated the pull request.

teohhanhui commented:
Anyway, please add some tests. :smile:

maechler commented:
I will add some tests, probably I'll need some time to have a look at the existing tests.

What about the docs, do you update them or shall I create a pull request there too?

teohhanhui commented on src/JsonLd/ContextBuilder.php:
```diff
@@ -97,18 +97,25 @@ public function getResourceContext(string $resourceClass, int $referenceType = U
                     '@type' => '@id',
-                ];
-            } else {
+                ] + $jsonldContext;
```
Reverse the order so that what's set in `jsonld_context` takes precedence.

maechler replied:
Ok, I thought in this case it might be better to force @id and @type. But the other way round it leaves more freedom for the developers. I'll change it.

maechler commented:
Do I have to take care of the deps='low' build in that case?

teohhanhui commented:
This should also address #731

@albertedevigo What do you think?

maechler commented:
@teohhanhui What about the unit tests?

teohhanhui commented:
You should add a test for the `JsonLdContextBuilder` too.

maechler commented:
@teohhanhui I added a test for `JsonLd/ContextBuilder`. 

I think #731 can be addressed with this issue. However I think there is also a need for a jsonld context on a resource level. I have created an issue here https://github.com/api-platform/api-platform/issues/93 and now moved to https://github.com/api-platform/core/issues/767.
That would also allow for the alternative use of `@reverse` described first in the issue.

dunglas commented:
Thank you @maechler! Can you open a doc PR on https://github.com/api-platform/docs/ to document this new feature please?

maechler commented:
@dunglas Awesome, thanks!
Yes I will. Shall I add a new .md file to the core doc and an own documentation chapter "jsonld context"? I have not found any chapter where it would make sense to add this.

dunglas commented:
I think that it deserves a new chapter.
